README
BOUNDS

BOUNDS is a command-line program for the GNU operating system that generates a boundary of a set of xy points.

With Bounds, generate an ogr-gmt compatible boundary vector of a given point set. There are a number of algorithms to choose from in generating the boundary, including convex hull, concave hull, bounding block and bounding box.

bounds infile [Options]

  Generate a boundary of the given xy point file(s) and print results 
  to stdout. Optionally, choose between different algorithms to use to 
  generate the boundary. The default behavior will generate a convex hull 
  boundary. 

  infile		The input point file.

 Options:
  -d, --delimiter	The input xy file record delimiter
     Note: The default will assume a space (" ") is the delimiter.
  -r, --record		The input record order, 'xy' should represent the locations
              		of the x and y records, respectively.
     Note: The default will assume the record begins with 'xy'. If the x and 
           y records are located elsewhere, fill in records with any letter. i.e.
           `-r zyx` for data where the record is `elevation,y-value,x-value`.
  -s, --skip		Skip header lines in the input xy file; specify the number of lines
            		to skip here.

  -b, --box		Generate a 'bounding box' boundary. 
  -k, --block		Generate a 'block' boundary. 
             		Specify the blocking increment here. (e.g. --block 0.001)
     Note: The blocking increment should be in the same units as the input xy data.
  -x, --convex		Generate a 'convex hull' boundary using a monotone chain algorithm. (Default)
  -v, --concave		Generate a 'concave hull' boundary using a distance weighted 
               		package wrap algorithm. 
               		Specify the distance threshold used in generating the concave 
               		hull as an argument here. (e.g. --concave 0.01)
     Note: The distance threshold should be in the same units as the input xy data.

  --help		Print this help menu and exit.
  --version		Print version information and exit.


NEWS
** 10/25/16 **

- New release (0.4.0)
- Added 'block_pts' boundary option for generating a polygon around a set of points by gridding the dataset and polygonizing the result.
- Bug Fixes in bounds.c and hull.c

**************

src/hull.c
/*------------------------------------------------------------
 * hull.c
 *
 * Copyright (c) 2011-2018 Matthew Love <matthew.love@colorado.edu>
 * This file is liscensed under the GPL v.2 or later and 
 * is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details. 
 * <http://www.gnu.org/licenses/> 
 *--------------------------------------------------------------*/

#include "bounds.h"

/* Get the euclidean distance between p1 and p2 in meters
 */
static float
dist_euclid(point_t* p1, point_t* p2) {
  return sqrt( ((p2->y - p1->y) * (p2->y - p1->y)) + ((p2->x - p1->x) * (p2->x - p1->x)) );
}

/* Calculate the angle between two points, from the previous point
 * p1 = current point, p2 = next point, p0 = previous point
 */
static float
atheta(point_t* p1, point_t* p2, point_t* p0) {
  float t, t2, t3, d12x, d12y, d13x, d13y;

  d12x = p1->x - p0->x, d12y = p1->y - p0->y; 
  d13x = p1->x - p2->x, d13y = p1->y - p2->y; 

  t = atan2(d12y, d12x), t2 = atan2(d13y, d13x);

  t3 = (t2 - t);

  if (t3 < 0) t3 = t3 + (2 * M_PI);
  return t3;
}

static float
atheta_degrees(point_t p1, point_t p2, point_t p0) {
  float t, t2, t3, d12x, d12y, d13x, d13y;

  d12x = p1.x - p0.x, d12y = p1.y - p0.y; 
  d13x = p1.x - p2.x, d13y = p1.y - p2.y; 

  t = atan2(d12y, d12x), t2 = atan2(d13y, d13x);

  t3 = (t2 - t) * 180 / M_PI;

  if (t3 < 0) t3 = t3 + 360;
  return t3;
}

/* Return the angle between the two given points 
 */
static float
theta(point_t* p1, point_t* p2) {
  float t = atan2(p2->y - p1->y, p2->x - p1->x);;
  if (t < 0) t = t + (2 * M_PI);
  return t;
}

static float
theta_degrees(point_t* p1, point_t* p2) {
  float t = atan2(p2->y - p1->y, p2->x - p1->x) * 180 / M_PI;
  if (t < 0) t = t + 360;
  return t;
}
 
/* Three points are a counter-clockwise turn if ccw > 0, clockwise if
 * ccw < 0, and collinear if ccw = 0 because ccw is a determinant that
 * gives the signed area of the triangle formed by p1, p2 and p3.
 */
static double
ccw(point_t* p1, point_t* p2, point_t* p3) {
  return (p2->x - p1->x) * (p3->y - p1->y) - (p2->y - p1->y) * (p3->x - p1->x);
}

int
intersect(line_t l1, line_t l2, double thresh) {
  double a, b;
  a = ccw(&l1.p1, &l1.p2, &l2.p1) *ccw(&l1.p1, &l1.p2, &l2.p2);
  b = ccw(&l2.p1, &l2.p2, &l1.p1) *ccw(&l2.p1, &l2.p2, &l1.p2);

  if (a < 0-thresh && b < 0-thresh) return 1;
  else if (a >= 0-thresh && a <= thresh) return 0;
  else if (b >= 0-thresh && b <= thresh) return 0;
  else return -1;
}

/* Depreciated intersect function. 
 */
static int
intersectp(point_t* p1, point_t* p2, point_t* p3, point_t* p4) {
  float den = (p2->x - p1->x) * (p4->y - p3->y) - (p2->y - p1->y) * (p4->x - p3->x);

  //if (fabs(den) < FLT_EPSILON) return 0; // parallel lines
  if (!den) return 0; // parallel lines

  float a = ((p3->x - p1->x) * (p4->y - p3->y) - (p3->y - p1->y) * (p4->x - p3->x)) / den;
  if (a <= FLT_EPSILON  || a >= 1.0) return 0; // intersection point not between p1 and p2

  a = ((p3->x - p1->x) * (p2->y - p1->y) - (p3->y - p1->y) * (p2->x - p1->x)) / den;
  if (a <= FLT_EPSILON || a >= 1.0) return 0; // intersection point not between p3 and p4

  return 1;
}

int
inside(point_t* p1, point_t* poly, ssize_t hullsize, double dist) {

  int k = 0, l, l1;
  ssize_t i;
  point_t p2;
  line_t lt, lp;

  p2.y = p1->y, p2.x = INT_MAX;
  lt.p1 = *p1, lt.p2 = p2;
  
  for (i = 0; i < hullsize; i++) {
    lp.p1 = poly[i], lp.p2 = poly[i+1];

    l = intersect(lt,lp,FLT_EPSILON*dist);

    if (l == 0) return 1;
    if (l == 1) k++;
  }
  return k & 1;
}

/* A 'package-wrap' convexhull 
 * -- Retruns the number of points in the boundary;
 * The hull makes up the begining of the points array.
 * Generate the concave hull using the given distance threshold
 */
ssize_t
dpw_concave(point_t* points, int npoints, double d) {
  int i, min, M, k, j, fp=0;
  float th, cth;
  point_t t, t0;
  line_t l1, l2;

  /* Find the minimum y value in the dataset */
  for (min = 0, i = 1; i < npoints-1; i++)
    if (points[i].y < points[min].y) min = i;

  /* Loop through all the points, starting with the point found above. */
  for (M = 0; M < npoints; M++) {
    t = points[M], points[M] = points[min], points[min] = t;
    min = -1, k = 0, th = 2*M_PI;

    /* Set the previous point for angle calculations */
    if (M > 0) t0 = points[M - 1];

    /* 
     *  !! Need a better way to re-insert first point !! 
     */
    /* Re-insert the first point into the dataset */
    if (M > 10 || M > npoints*.1) points[npoints-1] = points[0];
 
    /* Loop through the remaining points and find the next concave point; 
       less than distance threshold -> less than working theta -> does not 
       intersect existing boundary */
    for (i = M + 1; i < npoints; i++)
      if (dist_euclid(&points[M], &points[i]) <= d) {
	if (M == 0) cth = theta(&points[M], &points[i]);
	else cth = atheta(&points[M], &points[i], &t0);
	if (cth > 0  && cth < th) {
	  /* Make sure the selected point doesn't create a line segment which
	     intersects with the existing hull. */
	  if (M > 0)
	    for (k = 0, j = 0; j < M; j++) {
	      l1.p1 = points[M], l1.p2 = points[i];
	      l2.p1 = points[j], l2.p2 = points[j+1];
	      if (intersect(l1,l2,DBL_EPSILON) > 0) k++;
	    }
	  /* If all criteria are met,, add this point to the hull */
	  if (k == 0) min = i, th = cth; 
	}
      }

    /* No point was found, try again with a larger distance threshhold. */
    if (min == -1) return min;

    /* The first point was found again, a successful hull was found! end here. */
    if (min == npoints-1) {
      points[M + 1] = points[min];
      return M+1;
    }
  }
}

/* 
 * A Monotone-Chain Convex Hull
 * -- Returns a list of points on the convex hull in counter-clockwise order.
 * Note: the last point in the returned list is the same as the first one. 
 */
void
mc_convex(point_t* points, ssize_t npoints, point_ptr_t** out_hull, ssize_t* out_hullsize) {
  point_ptr_t* hull;
  ssize_t i, t, k = 0;

  hull = *out_hull;

  /* lower hull */
  for (i = 0; i < npoints; ++i) {
    while (k >= 2 && ccw(hull[k - 2], hull[k - 1], &points[i]) <= 0) --k;
    hull[k++] = &points[i];
  }
 
  /* upper hull */
  for (i = npoints - 2, t = k + 1; i >= 0; --i) {
    while (k >= t && ccw(hull[k - 2], hull[k - 1], &points[i]) <= 0) --k;
    hull[k++] = &points[i];
  }
  
  *out_hull = hull;
  *out_hullsize = k;
}

/* A 'package-wrap' Convex Hull 
 * -- Retruns the number of points in the hull;
 * The hull makes up the begining of the points array.
 */
ssize_t
pw_convex(point_t* points, ssize_t npoints) {
  ssize_t i, min, M;
  float th = 0.0;
  float cth, v;
  point_t t;

  for (min = 0, i = 1; i < npoints; i++)
    if (points[i].y < points[min].y) min = i;

  for (M = 0; M < npoints; M++) {
    t = points[M], points[M] = points[min]; 
    points[min] = t, min = npoints; 
    v = th, th = 2 * M_PI;

    for (i = M+1; i <= npoints; i++) {
      cth = theta(&points[M], &points[i]);
      if (cth > v)
	if (cth < th)
	  min = i, th = cth; 
    }
    if (min == npoints) {
      points[M + 1] = points[0];
      return M + 2;
    }
  }
}


src/pnts.c
/*------------------------------------------------------------
 * pnts.c
 *
 * Copyright (c) 2011-2018 Matthew Love <matthew.love@colorado.edu>
 * This file is liscensed under the GPL v.2 or later and 
 * is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details. 
 * <http://www.gnu.org/licenses/> 
 *--------------------------------------------------------------*/

#include "bounds.h"

/* Line-Count 
 */
ssize_t
linecnt(FILE *in) {
  char tmp[MAX_RECORD_LENGTH] = {0x0};
  int recordcnt = 0;

  while(fgets(tmp, sizeof(tmp), in) != 0) recordcnt++;
  rewind(in);

  return recordcnt; 
}

/* Scan an xyz line and attempt to determine the delimiter and record length 
 */
int
scanline(char* infile) {
  FILE *in = fopen(infile, "r");  
  int c, count;

  count = 0;
  for (;;) {
    c = fgetc(in);
    if (c == EOF || c == '\n') break;
    ++count;
  }
  return count;
}

/* Read a point record from the given infile
 */
int
read_point(FILE *infile, point_t *rpnt, char* delimiter, char* pnt_recr) {
  char tmp[MAX_RECORD_LENGTH] = {0x0};
  char pntp, *strs;

  int pf_length, j, status;

  /* Read in the file */
  if(infile == NULL) {
    perror("bounds: File open error");
    exit(EXIT_FAILURE);
  }

  /* read a record */
  if (fgets(tmp, sizeof(tmp), infile) !=0) {
    strs = malloc(strlen(tmp) + 1);
    strcpy(strs, tmp);
    status = 0;
  }
  else {
    status = -1;
    return status;
  }

  pf_length = strlen(pnt_recr);

  char* p = strtok(strs, delimiter);
  for (j = 0; j < pf_length; j++) {
    pntp = pnt_recr[j];
    if (p != NULL) {
      if (pntp == 'x') rpnt->x = atof(p);
      else if (pntp == 'y') rpnt->y = atof(p);
    }
    p = strtok(NULL, delimiter);
  }
  free(strs);  
  return status;
}

void
minmax(point_t* points, int npoints, xyz_info *xyzi) {
  int i;

  for (i = 0; i < npoints; i++) {
    if (i == 0) {
      xyzi->xmin = points[i].x, xyzi->ymin = points[i].y;
      xyzi->xmax = points[i].x, xyzi->ymax = points[i].y;
    }
    else {
      if (points[i].x <= xyzi->xmin) xyzi->xmin = points[i].x;
      if (points[i].x >= xyzi->xmax) xyzi->xmax = points[i].x;
      if (points[i].y <= xyzi->ymin) xyzi->ymin = points[i].y;
      if (points[i].y >= xyzi->ymax) xyzi->ymax = points[i].y;
    }
  }
}


src/block.c
/*------------------------------------------------------------
 * block.c
 *
 * Copyright (c) 2011-2018 Matthew Love <matthew.love@colorado.edu>
 * This file is liscensed under the GPL v.2 or later and 
 * is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details. 
 * <http://www.gnu.org/licenses/> 
 *--------------------------------------------------------------*/

#include "bounds.h"

/* 
 * "Bounding Block"
 * Generates a grid at `inc` cell-size and polygonizes it.
 */
int
block_pts(point_t* points, int npoints, double inc) {

  int i, j, xpos, ypos, edge;
  int bb = 0, done = 0, bcount = 0, l = 0, fcount = 0;
  double x, y;
  point_t bb1, bb2;

  /* Gather xyzinfo */
  xyz_info xyzi;
  minmax(points, npoints, &xyzi);

  /* Set the rows and columns of the internal grid */
  int ysize = fabs((xyzi.ymax - xyzi.ymin) / inc) + 1;
  int xsize = fabs((xyzi.xmax - xyzi.xmin) / inc) + 1;
  /* `xys` is the size of the bbarray array, this has to be large enough
   to hold all 4 sides of every point in the grid. */
  int xys = (xsize*ysize)*4;
  
  fprintf(stderr,"bounds: Size of internal grid: %d/%d\n",ysize,xsize);

  /* Allocate memory for arrays */  
  int** blockarray;
  blockarray = (int**) malloc(ysize * sizeof(int*));  
  for (i = 0; i < ysize; i++) blockarray[i] = (int*) malloc(xsize * sizeof(int));  
  
  if (!blockarray){
    fprintf(stderr,"bounds: Failed to allocate needed memory, try increasing the distance value (%f)\n",inc);
    exit(EXIT_FAILURE);
  }

  //fprintf(stderr,"bounds: Initializing bbarray with a value of %d\n",xys);
  line_t *bbarray;
  bbarray = (line_t *)malloc(sizeof(line_t)*xys);

  if (!bbarray) {
    fprintf(stderr,"bounds: Failed to allocate needed memory, try increasing the distance value (%f)\n",inc);
    exit(EXIT_FAILURE);
  }
  //fprintf(stderr,"bounds: bbarray Initialized\n");
  
  /* Loop through the point records and grid them */
  for (i = 0; i < npoints; i++) {
    xpos = fabs((points[i].x - xyzi.xmin) / inc);
    ypos = fabs((points[i].y - xyzi.ymin) / inc);
    blockarray[ypos][xpos] = 1;
  }
  
  /* Loop through the grid and find and record the edge lines */
  for (i = 0; i < ysize; i++) {
    for (j = 0; j < xsize; j++) {
      if (blockarray[i][j] == 1) {
	if (i == 0 || blockarray[i-1][j] != 1) {
	  bb1.x = ((j * inc) + xyzi.xmin), bb1.y = ((i * inc) + xyzi.ymin);
	  bb2.x = ((j * inc) + xyzi.xmin) + inc, bb2.y = ((i * inc) + xyzi.ymin);
	  bbarray[bb].p1 = bb1, bbarray[bb].p2 = bb2;
	  bb++;
	}
	if (j == 0 || blockarray[i][j-1] != 1) {
	  bb1.x = ((j * inc) + xyzi.xmin), bb1.y = ((i * inc) + xyzi.ymin);
	  bb2.x = ((j * inc) + xyzi.xmin), bb2.y = ((i * inc) + xyzi.ymin) + inc;
	  bbarray[bb].p1 = bb1, bbarray[bb].p2 = bb2;
	  bb++;
	}
	if (i == ysize-1 || blockarray[i+1][j] != 1) {
	  bb1.x = ((j * inc) + xyzi.xmin) + inc, bb1.y = ((i * inc) + xyzi.ymin) + inc;
	  bb2.x = ((j * inc) + xyzi.xmin), bb2.y = ((i * inc) + xyzi.ymin) + inc;
	  bbarray[bb].p1 = bb1, bbarray[bb].p2 = bb2;
	  bb++;
	}
	if (j == xsize-1 || blockarray[i][j+1] != 1) {
	  bb1.x = ((j * inc) + xyzi.xmin) + inc, bb1.y = ((i * inc) + xyzi.ymin);
	  bb2.x = ((j * inc) + xyzi.xmin) + inc, bb2.y = ((i * inc) + xyzi.ymin) + inc;
	  bbarray[bb].p1 = bb1, bbarray[bb].p2 = bb2;
	  bb++;
	}
      }
    }
  }

  /* Sort the edge lines into polygons */
  while (bb >= 4) {
    /* Do for each polygon...finishes (done == 1) when finds a matching point. */
    while (done == 0) {
      if (bcount == 0) {
	points[0] = bbarray[bb-1].p1, points[1] = bbarray[bb-1].p2;
	bcount=2, bb--;
      }
      for (edge = 0; edge < bb; edge++) {
	if (fabs( points[bcount-1].x - bbarray[edge].p1.x ) < FLT_EPSILON && fabs( points[bcount-1].y - bbarray[edge].p1.y) < FLT_EPSILON ) {
	  if (fabs( points[0].x - bbarray[edge].p2.x) < FLT_EPSILON && fabs (points[0].y - bbarray[edge].p2.y) < FLT_EPSILON) l = 1;
	  points[bcount] = bbarray[edge].p2;
	  bbarray[edge] = bbarray[bb-1];
	  bb--, bcount++, done = l;
	  break;
	}
	if (fabs (points[bcount-1].x - bbarray[edge].p2.x) < FLT_EPSILON && fabs (points[bcount-1].y - bbarray[edge].p2.y) < FLT_EPSILON) {
	  if (fabs (points[0].x - bbarray[edge].p1.x) < FLT_EPSILON && fabs (points[0].y - bbarray[edge].p1.y) < FLT_EPSILON) l = 1;
	  points[bcount] = bbarray[edge].p1;
	  bbarray[edge] = bbarray[bb-1];  
	  bb--, bcount++, done = l;
	  break;
	}
      }
    }

    /* Print out the polygon */
    for (edge = 0; edge < bcount; edge++) {
      printf("%f %f\n", points[edge].x, points[edge].y);
    }

    /* Reset some values */
    fcount = fcount + (bcount-1);
    bcount = 0, done = 0, l = 0;
    if (bb >= 4) printf(">\n");
  }
  
  fprintf(stderr,"bounds: Found %d total boundary points\n", fcount);

  /* Free allocated grid 'blockarray' from memory */
  for (i = 0; i < ysize; i++) free(blockarray[i]);
  free(blockarray);
  
  /* Free allocated grid bbarray from memory */
  free(bbarray);
  
  return(1);  
}


src/bounds.c
/*------------------------------------------------------------
 * bounds.c
 * 
 * Copyright (c) 2011-2018 Matthew Love <matthew.love@colorado.edu>
 * This file is liscensed under the GPL v.2 or later and 
 * is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details. 
 * <http://www.gnu.org/licenses/> 
 *--------------------------------------------------------------*/

#include <getopt.h>
#include "bounds.h"

/* Flags set by `--version' and `--help' */
static int version_flag;
static int help_flag;

static void
print_version(const char* command_name, const char* command_version) {
  fprintf(stderr, "%s version %s \n\
Copyright © 2011-2016 Matthew Love <matthew.love@colorado.edu> \n\
%s is liscensed under the GPL v.2 or later and is \n\
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n\
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A\n\
PARTICULAR PURPOSE.  See the GNU General Public License for more details.\n\
<http://www.gnu.org/licenses/>\n", command_name, command_version, command_name);
  exit (1);
}

static void 
usage() {
  fprintf(stderr, "\
bounds infile [Options]\n\n\
n  Generate a boundary of the given xy point file(s) and print results \n\
  to stdout. Optionally, choose between different algorithms to use to \n\
  generate the boundary. The default behavior will generate a convex hull \n\
  boundary. \n\
\n\
  infile\t\tThe input point file.\n\
\n\
 Options:\n\
  -d, --delimiter\tThe input xy file record delimiter\n\
     Note: The default will assume a space (\" \") is the delimiter.\n\
  -r, --record\t\tThe input record order, 'xy' should represent the locations\n\
              \t\tof the x and y records, respectively.\n\
     Note: The default will assume the record begins with 'xy'. If the x and \n\
           y records are located elsewhere, fill in records with any letter. i.e.\n\
           `-r zyx` for data where the record is `elevation,y-value,x-value`.\n\
  -s, --skip\t\tSkip header lines in the input xy file; specify the number of lines\n\
            \t\tto skip here.\n\n\
  -b, --box\t\tGenerate a 'bounding box' boundary. \n\
  -k, --block\t\tGenerate a 'block' boundary. \n\
             \t\tSpecify the blocking increment here. (e.g. --block 0.001)\n\
     Note: The blocking increment should be in the same units as the input xy data.\n\
  -x, --convex\t\tGenerate a 'convex hull' boundary using a monotone chain algorithm. (Default)\n\
  -v, --concave\t\tGenerate a 'concave hull' boundary using a distance weighted \n\
               \t\tpackage wrap algorithm. \n\
               \t\tSpecify the distance threshold used in generating the concave \n\
               \t\thull as an argument here. (e.g. --concave 0.01)\n\
     Note: The distance threshold should be in the same units as the input xy data.\n\
\n\
  --help\t\tPrint this help menu and exit.\n\
  --version\t\tPrint version information and exit.\n\n\
");
  exit(1);
}

/* Compare functions for use in qsort
 */
int
compare(const void* a, const void* b) {
  const point_t *elem1 = a;    
  const point_t *elem2 = b;

  if (elem1->x < elem2->x) return -1;
  else if (elem1->x > elem2->x) return 1;
  else return 0;
}

int
sort_min_y(const void* a, const void* b) {
  const point_t *elem1 = a;    
  const point_t *elem2 = b;

  if (elem1->y < elem2->y) return -1;
  else if (elem1->y > elem2->y) return 1;
  else return 0;
}

int
sort_max_y(const void* a, const void* b) {
  const point_t *elem1 = a;    
  const point_t *elem2 = b;

  if (elem1->y > elem2->y) return -1;
  else if (elem1->y < elem2->y) return 1;
  else return 0;
}

int
sort_max_x(const void* a, const void* b) {
  const point_t *elem1 = a;    
  const point_t *elem2 = b;

  if (elem1->x > elem2->x) return -1;
  else if (elem1->x < elem2->x) return 1;
  else return 0;
}

int
sort_min_x(const void* a, const void* b) {
  const point_t *elem1 = a;    
  const point_t *elem2 = b;

  if (elem1->x < elem2->x) return -1;
  else if (elem1->x > elem2->x) return 1;
  else return 0;
}

int
main (int argc, char **argv) {
  char* fn;
  FILE* fp;

  int c, i, status, min, j;
  int inflag = 0, vflag = 0, sflag = 0, pc = 0, sl = 0;
  int cflag = 0, kflag = 0, calg = 0, bflag = 0;
  double dist = 1;

  point_t rpnt, pnt;
  point_ptr_t hull0[MAX_HULLS];
  point_ptr_t* hull = hull0;
  ssize_t hullsize;

  char* delim = " ";
  char* ptrec = "xy";
  
  while (1) {
    static struct option long_options[] =
      {
	/* These options set a flag. */
	{"version", no_argument, &version_flag, 1},
	{"help", no_argument, &help_flag, 1},
	/* These options don't set a flag.
	   We distinguish them by their indices. */
	{"delimiter", required_argument, 0, 'd'},
	{"skip", required_argument, 0, 's'},
	{"record", required_argument, 0, 'r'},
	{"box", no_argument, 0, 'b'},
	{"block", required_argument, 0, 'k'},
	{"convex", no_argument, 0, 'x'},
	{"concave", required_argument, 0, 'v'},
	{0, 0, 0, 0}
      };
    /* getopt_long stores the option index here. */
    int option_index = 0;
    
    c = getopt_long (argc, argv, "d:r:s:bk:xv:",
		     long_options, &option_index);
    
    /* Detect the end of the options. */
    if (c == -1) break;
    
    switch (c) {
    case 0:
      /* If this option set a flag, do nothing else now. */
      if (long_options[option_index].flag != 0)
	break;
      printf ("option %s", long_options[option_index].name);
      if (optarg)
	printf (" with arg %s", optarg);
      printf ("\n");
      break;
    case 'd':
      delim = optarg;
      break;
    case 'r':
      ptrec = optarg;
      break;
    case 's':
      sflag++;
      sl = atoi(optarg);
      break;
    case 'b':
      bflag++;
      break;
    case 'k':
      kflag++;
      dist = atof(optarg);
      break;
    case 'x':
      cflag++;
      break;
    case 'v':
      vflag++;
      dist = atof(optarg);
      break;
      
    case '?':
      /* getopt_long already printed an error message. */
      break;
      
    default:
      abort ();
    }
  }

  if (version_flag) print_version("bounds", BOUNDS_VERSION);
  if (help_flag) usage();
  if (inflag == 0 && optind >= argc) usage ();

  fn = argv[optind];
  if (fn) inflag++;
  if (inflag > 0) {
    fp = fopen(fn, "r");
    
    if (!fp) {
      fprintf(stderr,"bounds: Failed to open file: %s\n", fn);
      exit(0);
    } else fprintf(stderr,"bounds: Working on file: %s ...\n", fn);

    ssize_t npr = linecnt(fp);
    fprintf(stderr,"( %d points )\n",npr);

    /* `sl` is a user-defined value of records to skip at
     * the beginning of the file. If `sflag` is set, reduce 
     * `npr` accordingly. */
    if (sflag) npr = npr - sl;

    if (npr == 0) {
      fclose(fp);
      exit(0);
    }
    
    /* Allocate memory for the `pnts` and `pnts2` array using the total number of points.
     * `pnts2` is only used by concave.
     */
    point_t* pnts;  
    pnts = (point_t*) malloc(npr * sizeof(point_t));

    if (!pnts) {
      fprintf(stderr,"bounds: Failed to allocate memory needed for point storage of %d points.\n", npr);
      exit(0);
    }

    
    point_t* pnts2;  
    pnts2 = (point_t*) malloc(npr * sizeof(point_t));  
        
    if (!pnts2) {
      fprintf(stderr,"bounds: Failed to allocate memory needed for point storage of %d points.\n", npr);
      exit(0);
    }
    
    /* Read through the point records and record them in `pnts`.
     */
    i = 0;
    while (read_point(fp, &rpnt, delim, ptrec) == 0) {
      if (sl > 0) sl--;
      else {
	pnts[i].x = rpnt.x, pnts[i].y = rpnt.y;
	i++;
      }
    }

    /* This is for the GMT compatibility. More can be done here.
     */
    printf("# @VGMT1.0 @GMULTIPOLYGON\n# @NName\n# @Tstring\n# FEATURE_DATA\n");
    printf(">\n# @D%s\n# @P\n", fn);
    
    /* The default is a convex hull -- `cflag` */
    if (cflag == 0 && vflag == 0 && bflag == 0 && kflag == 0) cflag++;
    
    /* 
     * Monotone Chain Convex Hull Algorithm - -*Default*-
     */
    if (cflag == 1) {
      qsort(pnts, npr, sizeof(point_t), compare);
      mc_convex(pnts, npr, &hull, &hullsize);
      
      for (i = 0; i < hullsize; i++)
	printf("%f %f\n", hull[i]->x, hull[i]->y);
      
      fprintf(stderr, "bounds: Found %d convex boundary points.\n", hullsize);
    }
    
    /* 
     * 'Package Wrap' Convex Hull Algorithm 
     */
    else if (cflag == 1 && calg == 2){
      hullsize = pw_convex(pnts, npr);
      
      for (i = 0; i < hullsize; i++)
	printf("%f %f\n", pnts[i].x, pnts[i].y);
      
      fprintf(stderr, "bounds: Found %d convex boundary points.\n", hullsize);
    }
    
    /* 
     * Concave Hull - distance weighted pacakage wrap algorithm 
     * Note: will return a single polygon containing all the given
     * points, using whatever distance value is needed to
     * accomplish that.
     */
    else if (vflag == 1) {
      /* The distance parameter can't be less than zero */
      if (dist > 0) hullsize = -1;
      else hullsize = 0;
      
      /* Keep a copy of the original point-set in `pts2` in-case
       * we need to re-run with a higher `dist` value. */
      memcpy(pnts2, pnts, sizeof(point_t)*npr);

      /* Find a boundary, inrease the distance variable until a boundary is found. */
      while (hullsize == -1) {
	qsort(pnts, npr, sizeof(point_t), sort_max_x);
	hullsize = dpw_concave(pnts, npr, dist);

	/* If a hull was found, check that it gathered all the points.
	 * If there are points still outside the boundary, increase the
	 * distance and retry.
	 */
	if (hullsize >= 0)
	for (i = hullsize; i < npr; i++)
	  if (!inside(&pnts[i], pnts, hullsize, dist)) {
	    hullsize = -1;
	    break;
	  }
	
	/* If a hull wasn't found, increase the `dist` and try again. */
	if (hullsize == -1) {
	  dist = dist + (dist * 0.1), pc++;
	  fprintf(stderr,"\rbounds: Increasing distance ( %f )",dist);
	  fflush(stderr);
	  memcpy(pnts, pnts2, sizeof(point_t)*npr);
	}
	
	/* In case something funky happens. */
	if (dist == INFINITY || dist == NAN || dist <= FLT_EPSILON) hullsize = 0;
      }

      /* Print out the hull */
      for (i = 0; i <= hullsize; i++)
      	printf("%.7f %.7f\n", pnts[i].x, pnts[i].y);
      
      if (pc > 0) fprintf(stderr,"\n");
      fprintf(stderr, "bounds: Found %d concave boundary points at a %.6f distance threshhold.\n", hullsize, dist);
    }
    
    /* 
     * Bounding Box - Generate a box around the given points.
     */
    else if (bflag == 1) {
      int xmin, xmax, ymin, ymax;
      
      /* Find the extent values in the dataset */
      for (ymin = 0, ymax = 0, xmin = 0, xmax = 0, i = 1; i < npr; i++) {
	if (pnts[i].y < pnts[ymin].y) ymin = i;
	if (pnts[i].x < pnts[xmin].x) xmin = i;
	if (pnts[i].y > pnts[ymax].y) ymax = i;
	if (pnts[i].x > pnts[xmax].x) xmax = i;
      }
      printf("%f %f\n", pnts[xmin].x, pnts[ymin].y);
      printf("%f %f\n", pnts[xmin].x, pnts[ymax].y);
      printf("%f %f\n", pnts[xmax].x, pnts[ymax].y);
      printf("%f %f\n", pnts[xmax].x, pnts[ymin].y);
      printf("%f %f\n", pnts[xmin].x, pnts[ymin].y);
    }

    /* 
     * Bounding Block - polygonize a grid of the points using `dist` cell-size
     */
    else if (kflag == 1) {
      /* The distance parameter can't be less than zero */
      if (dist > 0) block_pts(pnts, npr, dist);
    }
    free(pnts), free(pnts2), fclose(fp);
  }
  else usage ();
  exit(1);
}
